lom/src/plugins/plugins_common/plugins_common.go

```diff
@@ -8,63 +8,177 @@ package plugins_common
+        return "Disabled"
+    default:
+        return "Unknown stage"
+    }
+}
+
+const (
+    MAX_PLUGIN_RESPONSES_DEFAULT = 100 /* Max number of reesponses that plugin can send per
+       anamolykey during last MAX_PLUGIN_RESPONSES_WINDOW_TIMEOUT
+      before plugin manager mark it as disabled. Applicable for plugin's with timeout */
+    MAX_PLUGIN_RESPONSES_WINDOW_TIMEOUT_DEFAULT = 60 * time.Second /* Interval in which plugin can send
+       MAX_PLUGIN_RESPONSES_DEFAULT responses per anamoly key */
 )
 
-// sent from plugin to plugin manager via heartbeat channel
+/*
+ * sent from plugin to plugin manager via heartbeat channel
+ */
 type PluginHeartBeat struct {
     PluginName string
     EpochTime  int64
 }
 
-// sent from plugin to plugin manager as a responce to getPluginId()
+/*
+ * sent from plugin to plugin manager as a responce to getPluginId()
+ */
 type PluginId struct {
     Name    string
     Version string
 }
 
-// IPluginMetadata has common methods that are used by plugin manager to manage plugins. Data remain same for all plugins
+/*
+ * IPluginMetadata has common methods that are used by plugin manager to manage plugins. Data remain same for all plugins
+ */
 type IPluginMetadata interface {
     GetPluginStage() PluginStage
     SetPluginStage(stage PluginStage)
+    CheckMisbehavingPlugins(pluginKey string) bool
+}
+
+/*
+ * RollingWindow is used to keep track of response times of plugin requests
+ */
+type PluginResponseRollingWindow struct {
+    mu       sync.Mutex
+    response map[string][]time.Time // map of pluginname+Anamolykey to slice of response times
 }
 
-// Holds all data specific to plugin, run time info, etc
+/*
+ * Holds all data specific to plugin, run time info, etc
+ */
 type PluginMetadata struct {
     ActionCfg   *lomcommon.ActionCfg_t
     StartedTime time.Time
     Pluginstage PluginStage // indicate the current plugin stage
     PluginId
+    mu                          sync.Mutex // Mutex to synchronize access to PluginStage field
+    PluginResponseRollingWindow            // rolling window of response times
     // ... other common metadata fields
 }
 
+/*
+ * Get the current stage of plugin
+ */
 func (gpl *PluginMetadata) GetPluginStage() PluginStage {
+    gpl.mu.Lock()
+    defer gpl.mu.Unlock()
     return gpl.Pluginstage
 }
 
+/*
+ * Set the current stage of plugin
+ */
 func (gpl *PluginMetadata) SetPluginStage(stage PluginStage) {
+    gpl.mu.Lock()
+    defer gpl.mu.Unlock()
     gpl.Pluginstage = stage
 }
 
-// TODO: Goutham : Add more common methods here
+/*
+ * CheckMisbehavingPlugins maintains rolling window for sertain time. Checks if the rolling window for the given plugin+Anamolykey
+ * has reached a certain size within the specified time.
+ * It returns true if the window size has reached the limit, false otherwise.
+ */
+func (gpl *PluginMetadata) CheckMisbehavingPlugins(pluginKey string) bool {
+    gpl.PluginResponseRollingWindow.mu.Lock()
```

Comment:
A plugin == An action
The action can indeed return multiple different keys and this maintains timestamp per key.
Time update & check happens upon response. There is only one response at any one time.
Wonder why do we need mutex?

Reply:
Logically you are correct. But since requests are handled in goroutines, there is a small possibility that multiple requests(I mean when  responses are returned) from same action+anomalykey might be executing CheckMisbehavingPlugins.

Reply:
made the code changes to mark plugin request success after plugin manager sends data to engine. SO lock is not needed. Updated code.